app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/SenderDevice.kt

```diff
@@ -0,0 +1,116 @@
+ * on the sender device to transfer the file to the receiver device at the FILE_TRANSFER_PORT port.
+ *
+ * It takes in the uri of all the files to be shared. For each file uri, creates a new connection &
+ * copies all the bytes from input stream of the file to the output stream of the receiver device.
+ * Also changes the status of the corresponding FileItem on the list of files for transfer.
+ *
+ * A single task is used by the sender for the entire transfer
+ */
+private const val TIME_OUT = 15000
+
+internal class SenderDevice(
+  private val wifiDirectManager: WifiDirectManager,
+  activity: Activity
+) {
+  private val contentResolver: ContentResolver = activity.contentResolver
+  private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO
+  suspend fun send(fileItems: Array<out FileItem?>) = withContext(ioDispatcher) {
+    if (!delayForSlowReceiverDevicesToSetupServer()) {
+      return@withContext false
+    }
+    val hostAddress =
+      wifiDirectManager.getFileReceiverDeviceAddress().hostAddress
+    var isTransferErrorFree = true
+    var fileIndex = 0
+    while (fileIndex < fileItems.size && this.isActive) {
+      val fileItem = fileItems[fileIndex]
+      try {
+        Socket().use { socket ->
+          Log.d("gouri", "${Thread.currentThread().name} thread")
+          contentResolver.openInputStream(fileItem?.fileUri!!).use { fileInputStream ->
+            socket.bind(null)
+            socket.connect(
+              InetSocketAddress(hostAddress, WifiDirectManager.FILE_TRANSFER_PORT),
+              TIME_OUT
+            )
+            Log.d(TAG, "Sender socket connected to server - " + socket.isConnected)
+            publishProgress(fileIndex, FileItem.FileStatus.SENDING.ordinal)
+            val socketOutputStream = socket.getOutputStream()
+            copyToOutputStream(fileInputStream!!, socketOutputStream)
+            if (BuildConfig.DEBUG) Log.d(TAG, "Sender: Data written")
+            publishProgress(fileIndex, FileItem.FileStatus.SENT.ordinal)
+          }
+        }
+      } catch (e: IOException) {
+        Log.e(TAG, e.message)
+        e.printStackTrace()
+        isTransferErrorFree = false
+        publishProgress(fileIndex, FileItem.FileStatus.ERROR.ordinal)
+      }
+      fileIndex++
+    }
+    return@withContext isTransferErrorFree
+  }
+
+  private suspend fun publishProgress(vararg values: Int?) {
+    withContext(Dispatchers.Main) {
+      val fileIndex = values[0]
+      val fileStatus = values[1]
+      if (fileIndex != null && fileStatus != null) {
+        wifiDirectManager.changeStatus(fileIndex, FileItem.FileStatus.values()[fileStatus])
+      }
+    }
+  }
+
+  @SuppressWarnings("MagicNumber")
+  private suspend fun delayForSlowReceiverDevicesToSetupServer(): Boolean {
+    try { // Delay trying to connect with receiver, to allow slow receiver devices to setup server
+      delay(3000)
+    } catch (e: InterruptedException) {
```

Comment:
not an exception that can be thrown by delay, really this method can be entirely replace by just delaying, if the job has been cancelled then after it returns from delay it will throw a cancellation exception and stop doing stuff

Reply:
done